File: 2025/day_7/day_7.py
import os

file_path = os.path.join(os.path.dirname(__file__), "input_day_7.txt")
with open(file_path, "r") as file:
    input_ = file.read().split("\n")

beams = set()
beams.add(start_pos := input_[0].find("S"))
p1_count = 0

tls = [0] * len(input_[1])
tls[start_pos] = 1

for row in input_[1:]:
    for i in range(len(row)):
        if row[i] == "^":
            if i in beams:
                p1_count += 1
                beams.discard(i)
                beams.add(i + 1)
                beams.add(i - 1)
            if tls[i] != 0:
                tls[i - 1] += tls[i]
                tls[i + 1] += tls[i]
                tls[i] = 0

print("Part 1 Answer:", p1_count)
print("Part 2 Answer:", sum(tls))

File: 2025/day_7/day_7.cpp
#include <fstream>
#include <iostream>
#include <numeric>
#include <string>
#include <unordered_set>
#include <vector>

int main() {
  std::ifstream file("input_day_7.txt");
  if (!file.is_open()) {
    std::cerr << "Failed to open input\n";
    return 1;
  }

  std::vector<std::string> input_;
  std::string line;
  while (std::getline(file, line)) input_.push_back(line);

  auto start_pos = input_[0].find('S');

  std::unordered_set<int> beams;
  beams.insert(start_pos);
  int p1_count = 0;

  std::vector<long long> tls(input_[1].size(), 0);
  tls[start_pos] = 1;

  for (int r = 1; r < input_.size(); r++) {
    for (int c = 0; c < input_[r].size(); c++) {
      if (input_[r][c] == '^') {
        if (beams.count(c)) {
          p1_count++;
          beams.erase(c);
          beams.insert(c + 1);
          beams.insert(c - 1);
        }
        if (tls[c] != 0) {
          tls[c - 1] += tls[c];
          tls[c + 1] += tls[c];
          tls[c] = 0;
        }
      }
    }
  }

  std::cout << "Part 1 Answer: " << p1_count << "\n";
  auto p2_count = std::accumulate(tls.begin(), tls.end(), 0ULL);
  std::cout << "Part 2 Answer: " << p2_count << "\n";
}